openhcl/virt_mshv_vtl/src/processor/tdx/mod.rs

```diff
@@ -2096,65 +2139,66 @@ impl<T: CpuIo> EmulatorSupport for UhEmulationState<'_, '_, T, TdxBacked> {
         self.vp.partition.caps.vendor
     }
 
-    fn state(&mut self) -> Result<x86emu::CpuState, Self::Error> {
-        let cr0 = self.vp.backing.cr0.read(&self.vp.runner);
-        let efer = self.vp.backing.efer;
-        let cs = TdxExit(self.vp.runner.tdx_vp_enter_exit_info()).cs();
-        let enter_state = self.vp.runner.tdx_enter_guest_state();
+    fn gp_sign_extend(&mut self, reg: Register) -> i64 {
+        self.gp(reg) as i64
+    }
 
-        // TODO TDX: Only supports VTL0
-        Ok(x86emu::CpuState {
-            gps: enter_state.gps,
-            segs: [
-                self.vp
-                    .read_segment(GuestVtl::Vtl0, TdxSegmentReg::Es)
-                    .into(),
-                cs.into(),
-                self.vp
-                    .read_segment(GuestVtl::Vtl0, TdxSegmentReg::Ss)
-                    .into(),
-                self.vp
-                    .read_segment(GuestVtl::Vtl0, TdxSegmentReg::Ds)
-                    .into(),
-                self.vp
-                    .read_segment(GuestVtl::Vtl0, TdxSegmentReg::Fs)
-                    .into(),
-                self.vp
-                    .read_segment(GuestVtl::Vtl0, TdxSegmentReg::Gs)
-                    .into(),
-            ],
-            rip: enter_state.rip,
-            rflags: enter_state.rflags.into(),
-            cr0,
-            efer,
-        })
+    fn gp(&mut self, reg: Register) -> u64 {
+        let index = reg.number();
+        self.registers.gps[index]
     }
 
-    fn set_state(&mut self, state: x86emu::CpuState) -> Result<(), Self::Error> {
-        // TODO: immutable true? copied from snp
-        let x86emu::CpuState {
-            gps,
-            segs: _, // immutable
-            rip,
-            rflags,
-            cr0: _,  // immutable
-            efer: _, // immutable
-        } = state;
-        let enter_state = self.vp.runner.tdx_enter_guest_state_mut();
+    fn set_gp(&mut self, reg: Register, v: u64) {
+        let index = reg.number();
+        self.registers.gps[index] = v;
+    }
 
-        enter_state.gps = gps;
-        enter_state.rip = rip;
-        enter_state.rflags = rflags.into(); // TODO: rflags means interrupt state changed??
+    fn xmm(&mut self, index: usize) -> u128 {
+        u128::from_ne_bytes(self.vp.runner.fx_state().xmm[index])
+    }
+
+    fn set_xmm(&mut self, index: usize, v: u128) -> Result<(), Self::Error> {
```

Comment:
We should add some documentation on which registers get set immediately, and which only get set when you call flush.

Reply:
And which are immutable i guess.